app/controllers/products_controller.rb

```diff
@@ -18,6 +18,46 @@ def create
     end
   end
 
+  def show
+    @products = Product.all
+    @products.each do |pro|
+      if params[:id].to_i == pro.id
+        @product = Product.find(params[:id])
+      end
+    end
+    if @product == nil
+      redirect_to action: :index
+    end
+    @user_products = UserProduct.all
```

Comment:
```suggestion
@product = Product.find(params[:id])
```

大変申し訳ございません。こちら、上記のように修正くださいませ。
念の為、挙動確認も頂けますと幸いです。